app/uk/gov/hmrc/selfservicetimetopay/controllers/ArrangementController.scala

```diff
@@ -44,103 +44,96 @@ class ArrangementController(ddConnector: DirectDebitConnector,
-              Future.successful(Ok(instalment_plan_summary(ttp.schedule.getOrElse(throw new RuntimeException("No schedule data")),
-                createDayOfForm(ttp), signedIn = true)))
-            } else {
-              Future.successful(Redirect(routes.CalculatorController.getMisalignmentPage()))
-            }
-          })
-        }
+        case None => redirectToStart
+        case Some(ttp@TTPSubmission(Some(schedule), _, _, _, _, _, _, _, _)) if areEqual(ttp.taxpayer.get.selfAssessment.get.debits, ttp.calculatorData.debits) =>
+          Future.successful(Ok(instalment_plan_summary(schedule, createDayOfForm(ttp), signedIn = true)))
+        case Some(TTPSubmission(None, _, _, _, _, _, _, _, _)) => throw new RuntimeException("No schedule data")
+        case _ => Future.successful(Redirect(routes.CalculatorController.getMisalignmentPage()))
       }
   }
 
-  def submitInstalmentSummary: Action[AnyContent] = AuthorisedSaUser {
-    implicit authContext => implicit request =>
-      authorizedForSsttp {
-        Future.successful(Redirect(routes.DirectDebitController.getDirectDebit()))
-      }
+  def submitInstalmentSummary: Action[AnyContent] = AuthorisedSaUser { implicit authContext =>
+    implicit request =>
+      authorizedForSsttp(_ => Future.successful(Redirect(routes.DirectDebitController.getDirectDebit())))
   }
 
-  def changeSchedulePaymentDay(): Action[AnyContent] = AuthorisedSaUser {
-    implicit authContext => implicit request =>
-      authorizedForSsttp {
+  def changeSchedulePaymentDay(): Action[AnyContent] = AuthorisedSaUser { implicit authContext =>
+    implicit request =>
+      authorizedForSsttp { submission =>
         ArrangementForm.dayOfMonthForm.bindFromRequest().fold(
           formWithErrors => {
-            sessionCache.get.map {
-              submission => BadRequest(instalment_plan_summary(submission.get.schedule.get, formWithErrors, signedIn = true))
-            }
+            Future.successful(BadRequest(instalment_plan_summary(submission.get.schedule.get, formWithErrors, signedIn = true)))
           },
           validFormData => {
-            sessionCache.get.flatMap {
-              _.fold(redirectToStart)(ttp => changeScheduleDay(ttp, validFormData.dayOfMonth))
+            submission match {
+              case None => redirectToStart
+              case Some(ttp) => changeScheduleDay(ttp, validFormData.dayOfMonth).flatMap { ttpSubmission =>
+                sessionCache.put(ttpSubmission).map {
+                  _ => Redirect(routes.ArrangementController.getInstalmentSummary())
+                }
+              }
             }
           })
       }
   }
 
-  def changeScheduleDay(ttpSubmission: TTPSubmission, dayOfMonth: Int)(implicit hc: HeaderCarrier): Future[Result] = {
+  def changeScheduleDay(ttpSubmission: TTPSubmission, dayOfMonth: Int)(implicit hc: HeaderCarrier): Future[TTPSubmission] = {
```

Comment:
Amended function to NOT call the cache - the returned data is updated, the calling function should be responsible for determining what to do with it - aids in the "once per request" put to sessionCache